### PR TITLE
[WIP]*: use tidwall/btree replace of google/btree

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

ref https://github.com/pingcap/tidb/issues/34248

### CODE REVIEW

**Comment by sticnarf** [comment]
Sorry, client-go should guarantee it compiles on the oldest officially supported version, which is go 1.17 at this time.

If you hope to use generics here, you have to find some way (e.g. conditionally compilation) to let it fallback on the version which does not support generics.

**Comment by hawkingrei** [comment]
> Sorry, client-go should guarantee it compiles on the oldest officially supported version, which is go 1.17 at this time.
> 
> If you hope to use generics here, you have to find some way (e.g. conditionally compilation) to let it fallback on the version which does not support generics.



> Sorry, client-go should guarantee it compiles on the oldest officially supported version, which is go 1.17 at this time.
> 
> If you hope to use generics here, you have to find some way (e.g. conditionally compilation) to let it fallback on the version which does not support generics.

OK, I try to conditionally compilation solve it.

File: internal/locate/cache.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"bytes"
+	"sync"
+
+	"github.com/google/btree"
+)
+
+const btreeDegree = 32
+
+type CacheMu struct {
```
Comment (disksing) [inline internal/locate/cache.go:29]: Need some comments.

**Comment by siddontang** [comment]
what does CacheMu mean? 

please add tests for CacheMu

**Comment by hawkingrei** [comment]
To support old golang version, We need to use the build tag to support tidwall/btree and google/btree. it has many problems.

**Pull request closed without merging**